File: src/main/scala/com/wikibooks/spark/ch2/scala/AccumulatorSample.scala
package com.wikibooks.spark.ch2.scala

import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.util.AccumulatorV2

object AccumulatorSample {

  def runBuitInAcc(sc: SparkContext) {
    val acc1 = sc.longAccumulator("invalidFormat")
    val acc2 = sc.collectionAccumulator[String]("invalidFormat2")
    val data = List("U1:Addr1", "U2:Addr2", "U3", "U4:Addr4", "U5;Addr5", "U6:Addr6", "U7::Addr7")
    sc.parallelize(data, 3).foreach { v =>
      if (v.split(":").length != 2) {
        acc1.add(1L)
        acc2.add(v)
      }
    }
    println("잘못된 데이터 수:" + acc1.value)
    println("잘못된 데이터:" + acc2.value)
  }

  def runCustomAcc(sc: SparkContext) {
    val acc = new RecordAccumulator
    sc.register(acc, "invalidFormat")
    val data = List("U1:Addr1", "U2:Addr2", "U3", "U4:Addr4", "U5;Addr5", "U6:Addr6", "U7::Addr7")
    sc.parallelize(data, 2).foreach { v =>
      if (v.split(":").length != 2) {
        acc.add(Record(1))
      }
    }
    println("잘못된 데이터 수:" + acc.value)
  }

  def main(args: Array[String]) = {
    val conf = new SparkConf()
    conf.setMaster("local[*]").setAppName("AccumulatorSample")
    val sc = new SparkContext(conf)
    runCustomAcc(sc)
    sc.stop
  }
}

class RecordAccumulator extends AccumulatorV2[Record, Long] {

  private var _record = Record(0)

  def isZero: Boolean = _record.amount == 0 && _record.number == 1

  def copy(): AccumulatorV2[Record, Long] = {
    val newAcc = new RecordAccumulator
    newAcc._record = Record(_record.amount, _record.number)
    newAcc
  }

  def reset(): Unit = {
    _record.amount = 0L
    _record.number = 1L
  }

  def add(other: Record): Unit = {
    _record.add(other)
  }

  def merge(other: AccumulatorV2[Record, Long]): Unit = other match {
    case o: RecordAccumulator => _record.add(o._record);
    case _                    => throw new RuntimeException
  }

  def value: Long = {
    _record.amount
  }
}

File: src/main/java/com/wikibooks/spark/ch2/AccumulatorSample.java
package com.wikibooks.spark.ch2;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;
import org.apache.spark.util.AccumulatorV2;
import org.apache.spark.util.CollectionAccumulator;
import org.apache.spark.util.LongAccumulator;

import java.util.Arrays;
import java.util.List;

public class AccumulatorSample {

  // ex 2-141
  public static void runBuitInAcc(JavaSparkContext jsc) {
    LongAccumulator acc1 = jsc.sc().longAccumulator("invalidFormat");
    CollectionAccumulator acc2 = jsc.sc().collectionAccumulator("invalidFormat2");
    List<String> data = Arrays.asList("U1:Addr1", "U2:Addr2", "U3", "U4:Addr4", "U5;Addr5", "U6:Addr6", "U7::Addr7");
    jsc.parallelize(data, 3).foreach(new VoidFunction<String>() {
      @Override
      public void call(String v) throws Exception {
        if (v.split(":").length != 2) {
          acc1.add(1L);
          acc2.add(v);
        }
      }
    });
    System.out.println("잘못된 데이터 수:" + acc1.value());
    System.out.println("잘못된 데이터:" + acc2.value());
  }

  // ex 2-144
  public static void runCustomAcc(JavaSparkContext jsc) {
    RecordAccumulator acc = new RecordAccumulator();
    jsc.sc().register(acc, "invalidFormat");
    List<String> data = Arrays.asList("U1:Addr1", "U2:Addr2", "U3", "U4:Addr4", "U5;Addr5", "U6:Addr6", "U7::Addr7");
    jsc.parallelize(data, 3).foreach(new VoidFunction<String>() {
      @Override
      public void call(String v) throws Exception {
        if (v.split(":").length != 2) {
          acc.add(new Record(1L));
        }
      }
    });
    System.out.println("잘못된 데이터 수:" + acc.value());
  }

  public static void main(String[] args) throws Exception {
    SparkConf conf = new SparkConf().setAppName("AccumulatorSample").setMaster("local[*]");
    JavaSparkContext sc = new JavaSparkContext(conf);
    runCustomAcc(sc);
    sc.stop();
  }
}

class RecordAccumulator extends AccumulatorV2<Record, Long> {

  private static final long serialVersionUID = 1L;

  private Record _record = new Record(0L);

  @Override
  public boolean isZero() {
    return _record.amount == 0L && _record.number == 1L;
  }

  @Override
  public AccumulatorV2<Record, Long> copy() {
    RecordAccumulator newAcc = new RecordAccumulator();
    newAcc._record = new Record(_record.amount, _record.number);
    return newAcc;
  }

  @Override
  public void reset() {
    _record.amount = 0L;
    _record.number = 1L;
  }

  @Override
  public void add(Record other) {
    _record.add(other);
  }

  @Override
  public void merge(AccumulatorV2<Record, Long> otherAcc) {
    try {
      Record other = ((RecordAccumulator) otherAcc)._record;
      _record.add(other);
    } catch (Exception e) {
      throw new RuntimeException();
    }
  }

  @Override
  public Long value() {
    return _record.amount;
  }
}
